vllm/distributed/kv_transfer/kv_pipe/p2p_nccl_pipe.py

```diff
@@ -0,0 +1,287 @@
+    NCCLLibrary, buffer_type, cudaStream_t, ncclComm_t, ncclDataTypeEnum)
+from vllm.utils import current_stream, get_ip
+
+logger = logging.getLogger(__name__)
+
+
+class P2pNcclPipe:
+
+    def __init__(self,
+                 local_rank: int,
+                 config: KVTransferConfig,
+                 hostname: str = "",
+                 port_offset: int = 0,
+                 library_path: Optional[str] = None) -> None:
+        self.config = config
+        self.local_rank = local_rank
+        self.device = torch.device(f"cuda:{self.local_rank}")
+        self.nccl = NCCLLibrary(library_path)
+
+        if not hostname:
+            hostname = get_ip()
+        port = self.config.kv_port + port_offset
+        if port == 0:
+            raise ValueError("Port cannot be 0")
+        self._hostname = hostname
+        self._port = port
+        self.zmq_address = f"{self._hostname}:{self._port}"
+        self.http_address = (
+            f"{self._hostname}:"
+            f"{self.config.kv_connector_extra_config['http_port']}")
+        self.proxy_address = (
+            f"{self.config.kv_connector_extra_config['proxy_ip']}:"
+            f"{self.config.kv_connector_extra_config['proxy_port']}")
+        self.context = zmq.Context()
+        self.router_socket = self.context.socket(zmq.ROUTER)
+        self.router_socket.bind(f"tcp://{self.zmq_address}")
+
+        self.poller = zmq.Poller()
+        self.poller.register(self.router_socket, zmq.POLLIN)
+
+        self.send_store: Deque[List[Any]] = deque()  # tensor_id: torch.Tensor
+        self.recv_store: Dict[str,
+                              torch.Tensor] = {}  # tensor_id: torch.Tensor
+        self.socks: Dict[str, Any] = {}  # remote_address: client socket
+        self.comms: Dict[str, Any] = {}  # remote_address: (ncclComm_t, rank)
+
+        # self.buffer_size = 0
+        # self.buffer_size_threshold =  self.config.kv_buffer_size
+
+        self.send_store_cv = threading.Condition()
+        self.recv_store_cv = threading.Condition()
+        self.comm_cv = threading.Condition()
+
+        self._listener_thread = threading.Thread(
+            target=self._listen_for_requests, daemon=True)
+        self._listener_thread.start()
+
+        self._send_thread = threading.Thread(target=self._send_sync,
+                                             daemon=True)
+        self._send_thread.start()
+
+        if port_offset == 0:
+            self._ping_thread = threading.Thread(target=self._ping,
+                                                 daemon=True)
+            self._ping_thread.start()
+
+    def _create_connect(self, remote_address: typing.Optional[str] = None):
+        assert remote_address is not None
+        if remote_address not in self.socks:
+            sock = self.context.socket(zmq.DEALER)
+            sock.setsockopt_string(zmq.IDENTITY, self.zmq_address)
+            sock.connect(f"tcp://{remote_address}")
```

Comment:
For this PR, can we leverage `ipc`?

In general, using zmq sockets with pickle is insecure (see below)
- https://pyzmq.readthedocs.io/en/stable/api/zmq.html#zmq.Socket.recv_pyobj

With TCP, any arbitrary user with access to the route can cause remote code execution.

For this prototyping stage, we should only have ipc.